snuba/migrations/snuba_migrations/metrics/0002_metrics_sets.py

```diff
@@ -0,0 +1,160 @@
+    Column("granularity", UInt(32)),
+    Column("timestamp", DateTime()),
+    Column("retention_days", UInt(16)),
+    Column("tags", Nested([Column("key", UInt(64)), Column("value", UInt(64))]),),
+    Column("value", AggregateFunction("uniqCombined64", [UInt(64)])),
+]
+
+
+matview_statement = """
+SELECT
+    org_id,
+    project_id,
+    metric_id,
+    60 as granularity,
+    toStartOfMinute(timestamp) as timestamp,
+    retention_days,
+    tags.key,
+    tags.value,
+    uniqCombined64State(arrayJoin(set_values)) as value
+FROM metrics_buckets_local
+WHERE metric_type = 'set' AND materialization_version = 0
+GROUP BY
+    org_id,
+    project_id,
+    metric_id,
+    tags.key,
+    tags.value,
+    timestamp,
+    granularity,
+    retention_days
+"""
+
+
+class Migration(migration.ClickhouseNodeMigration):
+    blocking = False
+
+    def forwards_local(self) -> Sequence[operations.SqlOperation]:
+        return [
+            operations.CreateTable(
+                storage_set=StorageSetKey.METRICS,
+                table_name="metrics_sets_local",
+                columns=sets_columns,
+                engine=table_engines.AggregatingMergeTree(
+                    storage_set=StorageSetKey.METRICS,
+                    order_by="(org_id, project_id, metric_id, granularity, timestamp)",
+                    partition_by="(retention_days, toMonday(timestamp))",
+                    settings={"index_granularity": "256"},
+                ),
+            ),
+            # TODO: Fix the formatting we do in the migrations framework to be able
+            # to add columns called `tags.key` in the CREATE TABLE statement
+            # tags.key and tags.value have to be in the sorting key in order to be
+            # part of the primary key so that data from different tags does not
+            # get aggregated together.
+            operations.RunSql(
+                storage_set=StorageSetKey.METRICS,
+                statement="""
+                    ALTER TABLE metrics_sets_local ADD COLUMN IF NOT EXISTS `tags.key` Array(UInt64) AFTER retention_days,
+                    MODIFY ORDER BY (org_id, project_id, metric_id, granularity, timestamp, tags.key);
+                """,
+            ),
+            operations.RunSql(
+                storage_set=StorageSetKey.METRICS,
+                statement="""
+                    ALTER TABLE metrics_sets_local ADD COLUMN IF NOT EXISTS `tags.value` Array(UInt64) AFTER tags.key,
+                    MODIFY ORDER BY (org_id, project_id, metric_id, granularity, timestamp, tags.key, tags.value);
+                """,
+            ),
+            operations.AddColumn(
+                storage_set=StorageSetKey.METRICS,
+                table_name="metrics_sets_local",
+                column=Column(
+                    "_tags_hash",
+                    Array(UInt(64), Modifiers(materialized=INT_TAGS_HASH_MAP_COLUMN)),
+                ),
+                after="tags.value",
+            ),
```

Comment:
This nonsense is needed because you cannot create a nested column and reference it in the order by, but we cannot create the individual columns (tas.key, tags.value) because clickhouse wants those columns escaped, but the migraiton framework double escapes them.
I will see if I can fix the escaping.

Reply:
I thought these extra steps were because you wanted the table to have a different primary key to order by, and not specifically to do with escaping things?

Reply:
No, it is because if I create a `tags.key` column in the create statement, clickhouse does not like it.
If i make it a nested column clickhouse likes it ... until it gets to the order by, when it changes its mind
creating it as 
```
`tags.key` 
```
Works but our escaping logic double escapes it.

Reply:
I'm kind of confused about what the issue is here because the entire order by statement is just provided as a big string, and we don't attempt to escape individual terms within it at all. If you set the orderby in the create table statement to the below doesn't it just work?
`"(org_id, project_id, metric_id, granularity, timestamp, tags.key, tags.value)"`

Reply:
I cannot do this in clickhouse 
```
CREATE TABLE ....
    col.  type,
    tags.key. Array(string),
    ...
)
```
Clickhosue would tell me to use a nested column.

But the nested column would not be recognized by the ORDER BY. So I would have to format the create this way:
```
CREATE TABLE ....
    col.  type,
    `tags.key` Array(string),
    ...
)
```

But our migraiton code would escape that into:

```
CREATE TABLE ....
    col.  type,
   `\`tags.key\`` Array(string),
    ...
)
```

So I need to do it in multiple steps.

Reply:
Ok, for some reason I cannot reproduce the first issue anymore. 
Removed this complexity